spatialmath/geometry_utils.go

```diff
@@ -50,30 +52,72 @@ func SegmentDistanceToSegment(ap1, ap2, bp1, bp2 r3.Vector) float64 {
 }
 
 // ClosestPointsSegmentSegment will return the points at which two line segments are closest to one another.
-func ClosestPointsSegmentSegment(ap1, ap2, bp1, bp2 r3.Vector) (r3.Vector, r3.Vector) {
-	// vectors between line endpoints:
-	v0 := bp1.Sub(ap1)
-	v1 := bp2.Sub(ap1)
-	v2 := bp1.Sub(ap2)
-	v3 := bp2.Sub(ap2)
-
-	// squared distances:
-	d0 := v0.Norm2()
-	d1 := v1.Norm2()
-	d2 := v2.Norm2()
-	d3 := v3.Norm2()
-
-	// select best potential endpoint on capsule A:
-	bestA := ap1
-	if d2 < d0 || d2 < d1 || d3 < d0 || d3 < d1 {
-		bestA = ap2
-	}
-	// select point on capsule B line segment nearest to best potential endpoint on A capsule:
-	bestB := ClosestPointSegmentPoint(bp1, bp2, bestA)
-
-	// now do the same for capsule A segment:
-	bestA = ClosestPointSegmentPoint(ap1, ap2, bestB)
-	return bestA, bestB
+func ClosestPointsSegmentSegment(a1, a2, b1, b2 r3.Vector) (r3.Vector, r3.Vector) {
+	// Proceed according to this math.stackexchange answer: https://math.stackexchange.com/a/2812513
+	// The stack solution has a sign error, which we correct
+	// These (ugly) variable names were chosen to match those in the stack exchange answer, for easy reference
+
+	d3121 := b1.Sub(a1).Dot(a2.Sub(a1))
+	d4321 := b2.Sub(b1).Dot(a2.Sub(a1))
+	d4331 := b2.Sub(b1).Dot(b1.Sub(a1))
+	r1pow2 := a2.Sub(a1).Norm2()
+	r2pow2 := b2.Sub(b1).Norm2()
+
+	denom := r1pow2*r2pow2 - d4321*d4321
+	// If denom is 0, the segments are parallel, and we can jump to the endpt case below
+	// If denom is not 0, we finish our algebra
+	if denom != 0 {
```

Comment:
Thanks to floating point numbers, the chances of `denom` being precisely zero is extremely low in all cases except ones that are hand-constructed to turn out that way; we should check if the absolute value is smaller than `floatEpsilon` which currently is 1e-6. I'm open to a smaller value.